src/slip0010/ckd.rs

```diff
@@ -0,0 +1,141 @@
+//! Utilities for Child Key Derivation (CKD) according to SLIP-0010.
+
+use crate::{errors::CallerError, utils::bn_to_scalar};
+use generic_array::{
+    typenum::{U32, U64},
+    GenericArray,
+};
+use hmac::Mac;
+use k256::{ecdsa::VerifyingKey, Scalar};
+use libpaillier::unknown_order::BigNumber;
+
+type HmacSha512 = hmac::Hmac<sha2::Sha512>;
+
+/// Represents the input to the CKD function.
+#[derive(Debug, Clone)]
+pub struct CKDInput {
+    public_key: VerifyingKey,
+    chain_code: [u8; 32],
+    index: u32,
+}
+
+impl CKDInput {
+    /// Create a new CKDInput.
+    pub fn new(
+        public_key: VerifyingKey,
+        chain_code: [u8; 32],
+        index: u32,
+    ) -> Result<Self, CallerError> {
+        if index == 0 || index >= 0x80000000 {
+            Err(CallerError::WrongIndex)
+        } else {
+            Ok(Self {
+                public_key,
+                chain_code,
+                index,
+            })
+        }
+    }
+
+    /// Get the public key.
+    pub fn public_key(&self) -> VerifyingKey {
+        self.public_key
+    }
+
+    /// Get the chain code.
+    pub fn chain_code(&self) -> &[u8; 32] {
+        &self.chain_code
+    }
+
+    /// Get the index.
+    pub fn index(&self) -> u32 {
+        self.index
+    }
+
+    /// Derive master key from seed.
+    pub fn derive_master_key(seed: &[u8]) -> (Scalar, [u8; 32]) {
+        let hmac = HmacSha512::new_from_slice("Bitcoin seed".as_bytes())
+            .expect("this never fails: hmac can handle keys of any size");
+        let i = hmac.clone().chain_update(seed).finalize().into_bytes();
+        let (i_left, i_right) = Self::split_into_two_halfes(&i);
```

Comment:
I know we didn't ask for this, but because the SLIP10 codebase also does retries and checks if the seed generation fails, do you mind adding that in here as well? 

<img width="745" alt="Screenshot 2024-10-06 at 1 33 32 PM" src="https://github.com/user-attachments/assets/3e4080a4-94b6-444f-bda2-28e3bac79fe0">